### DISCUSSION TITLE
Each tenant's databases hosted by it's own server

### DISCUSSION THREAD

**Discussion by robingiraud** [spatie/laravel-multitenancy#458] [Q&A]
Hi everyone,

By default, this package's multi-tenant system assumes that all databases are hosted on the same server, as there is only the "database" field that is filled in the tenants table.
I need to be able to define a **different server for each tenant**, which would each host its own database.

So I started by modifying the migration that creates the tenants table, adding some extra fields: "host", "username" and "password".
I guess the next step is to create my own task that implements "SwitchTenantTask" interface, to define my custom way to set the current tenant's database connection, considering, in addition to the database name, the hostname, username and password to access the specific database. But I find it a bit tricky, as I need to override all the default logic provided by the package.

Is there a simpliest way to get there ? Or perhaps one of you had a similar need

In fact, to make it simple, I would like to take into account the host, username and password in the connection configuration of the current tenant (in addition to the database name)

Thanks a lot for any answer ! :)

**Comment by ChristiaanBilthof** [comment]
I had this requirement as well. Just extend their task and write / copy what you need yourself. You'll be done in a few seconds. If you need help, let me know.

**Reply by robingiraud** [reply]
Thanks for your help,

At first sight, I would need to modify this parts of the class:

- In the makeCurrent() function, pass all the $tenant as parameter instead of the database name only
- Rename the function from setTenantConnectionDatabaseName to setTenantConnectionDatabase (to be cleaner)
- In the setTenantConnectionDatabase function, add into the config array the extra fields:
  "database.connections.{$tenantConnectionName}.host" => $tenant->host, etc...
- In the app('db')->extend callback function, also add the extra field to change the config

Am I on the right track? Did you have any changes to make in your case?

Thanks,

**Reply by ChristiaanBilthof** [reply]
That's exactly what I've done haha.

**Reply by ChristiaanBilthof** [reply]
Btw, I said 'extend their task', but I just created my own and implemented the interface.

**Reply by m7vm7v** [reply]
I would recommend the use of password encryption for the databases as a safe step.

**Reply by robingiraud** [reply]
Thanks guys,

It seems to work well !

I have one last problem, which is that I get a 401 after I change my tenant. I have to clear my cookies before switching, and I don't remember having this issue before customizing the interface
For example I connect to tenant1.myapp.com correctly. Then I switch to tenant2.myapp.com but I have a 401, instead of landing on the login page.

Anyway, thanks again
